File: plugins/IntentionPowerPak/src/com/siyeh/ipp/concatenation/CopyConcatenatedStringToClipboardIntention.java
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.siyeh.ipp.concatenation;

import com.intellij.codeInspection.ModCommands;
import com.intellij.modcommand.ModCommand;
import com.intellij.modcommand.PsiBasedModCommandAction;
import com.intellij.psi.*;
import com.intellij.psi.util.PsiTreeUtil;
import com.siyeh.IntentionPowerPackBundle;
import com.siyeh.ig.psiutils.ExpressionUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author Bas Leijdekkers
 */
public class CopyConcatenatedStringToClipboardIntention extends PsiBasedModCommandAction<PsiExpression> {
  public CopyConcatenatedStringToClipboardIntention() {
    super(PsiExpression.class);
  }
  
  @Override
  public @NotNull String getFamilyName() {
    return IntentionPowerPackBundle.message("copy.concatenated.string.to.clipboard.intention.family.name");
  }

  @Override
  protected @Nullable Presentation getPresentation(@NotNull ActionContext context, @NotNull PsiExpression element) {
    final boolean isStringLiteral =
      element instanceof PsiLiteralExpression &&
      ExpressionUtils.hasStringType(element) &&
      !(PsiTreeUtil.skipParentsOfType(element, PsiParenthesizedExpression.class) instanceof PsiPolyadicExpression);
    if (isStringLiteral) {
      return Presentation.of(IntentionPowerPackBundle.message("copy.string.literal.to.clipboard.intention.name"));
    }
    if (ExpressionUtils.isStringConcatenation(PsiTreeUtil.getNonStrictParentOfType(element, PsiPolyadicExpression.class))) {
      return Presentation.of(IntentionPowerPackBundle.message("copy.concatenated.string.to.clipboard.intention.name"));
    }
    return null;
  }

  @Override
  protected @NotNull ModCommand perform(@NotNull ActionContext context, @NotNull PsiExpression element) {
    final String text;
    if (element instanceof PsiLiteralExpression literalExpression) {
      if (!(literalExpression.getValue() instanceof String string)) {
        return ModCommands.nop();
      }
      text = string;
    }
    else {
      PsiPolyadicExpression polyadic = PsiTreeUtil.getNonStrictParentOfType(element, PsiPolyadicExpression.class);
      if (!ExpressionUtils.isStringConcatenation(polyadic)) {
        return ModCommands.nop();
      }
      text = buildConcatenationText(polyadic);
    }
    return ModCommands.copyToClipboard(text);
  }

  public static String buildConcatenationText(PsiPolyadicExpression polyadicExpression) {
    final StringBuilder out = new StringBuilder();
    for(PsiElement element = polyadicExpression.getFirstChild(); element != null; element = element.getNextSibling()) {
      if (element instanceof PsiExpression expression) {
        final Object value = ExpressionUtils.computeConstantExpression(expression);
        out.append((value == null) ? "?" : value.toString());
      }
      else if (element instanceof PsiWhiteSpace && element.getText().contains("\n") &&
               (out.isEmpty() || out.charAt(out.length() - 1) != '\n')) {
        out.append('\n');
      }
    }
    return out.toString();
  }
}

File: plugins/kotlin/code-insight/intentions-k2/src/org/jetbrains/kotlin/idea/codeinsight/intentions/copyConcatenatedStringToClipboard/CopyConcatenatedStringToClipboardIntention.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.

package org.jetbrains.kotlin.idea.codeinsight.intentions.copyConcatenatedStringToClipboard

import com.intellij.codeInspection.util.IntentionFamilyName
import com.intellij.modcommand.ActionContext
import com.intellij.modcommand.ModCommand
import com.intellij.modcommand.Presentation
import com.intellij.modcommand.PsiBasedModCommandAction
import org.jetbrains.kotlin.analysis.api.analyze
import org.jetbrains.kotlin.idea.base.resources.KotlinBundle
import org.jetbrains.kotlin.lexer.KtTokens
import org.jetbrains.kotlin.psi.KtBinaryExpression

internal class CopyConcatenatedStringToClipboardIntention :
    PsiBasedModCommandAction<KtBinaryExpression>(KtBinaryExpression::class.java) {

    override fun getFamilyName(): @IntentionFamilyName String =
        KotlinBundle.message("copy.concatenation.text.to.clipboard")

    override fun getPresentation(context: ActionContext, element: KtBinaryExpression): Presentation? {
        if (element.operationToken != KtTokens.PLUS) return null

        val isString = analyze(element) {
            element.expressionType?.isStringType == true
        }
        if (!isString) return null

        return super.getPresentation(context, element)
    }

    override fun perform(context: ActionContext, element: KtBinaryExpression): ModCommand {
        val text = ConcatenatedStringGenerator().create(element)
        return ModCommand.copyToClipboard(text)
    }
}
